pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,197 @@
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		node.Spec.Taints = removeTaint(node.Spec.Taints, taintIndex)
+	}
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func removeTaint(s []kube_core.Taint, index int) []kube_core.Taint {
+	return append(s[:index], s[index+1:]...)
+}
+
+func (r *CniNodeTaintReconciler) taintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, kube_core.Taint{
+		Key:    nodeReadinessTaintKey,
+		Effect: kube_core.TaintEffectNoSchedule,
+	})
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("added taint to node")
+	}
+	return err
+}
+
+func hasTaint(node *kube_core.Node) bool {
+	foundTaint := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule {
+			foundTaint = true
+		}
+	}
+	return foundTaint
+}
+
+func hasCniPodRunning(log logr.Logger, pods []kube_core.Pod) bool {
+	for _, pod := range pods {
+		if isCniPod(pod) && pod.Status.Phase == kube_core.PodRunning {
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == kube_core.PodReady && condition.Status == kube_core.ConditionTrue {
+					log.V(1).Info("pod has kuma-cni-node running and ready", "pod", pod.Name, "status", pod.Status)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func isCniPod(pod kube_core.Pod) bool {
+	value, ok := pod.Labels[parentAppLabel]
+	return ok && value == metadata.KumaCNI
+}
+
+func (r *CniNodeTaintReconciler) SetupWithManager(mgr kube_ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &kube_core.Pod{}, nodeIndexField, func(obj kube_client.Object) []string {
+		pod := obj.(*kube_core.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
+	return kube_ctrl.NewControllerManagedBy(mgr).
+		For(&kube_core.Node{}, builder.WithPredicates(nodeEvents)).
+		// check this is necessary
+		Watches(
+			&kube_source.Kind{Type: &kube_core.Pod{}},
+			kube_handler.EnqueueRequestsFromMapFunc(podToNodeMapper(r.Log)),
+		).
+		Complete(r)
+}
+
+func podToNodeMapper(log logr.Logger) kube_handler.MapFunc {
+	return func(obj kube_client.Object) []kube_reconcile.Request {
+		pod, ok := obj.(*kube_core.Pod)
+		if !ok {
+			log.WithValues("pod", obj.GetName()).Error(errors.Errorf("wrong argument type: expected %T, got %T", pod, obj), "wrong argument type")
+			return nil
+		}
+
+		if pod.Spec.NodeName == "" {
```

Comment:
For some reason in the logs I was seeing a lot of `could not find a node with name ""` so this is why I'm filtering it out here. If you know why that might be let me know.

Reply:
Should this be a comment in the code?